### PR TITLE
[oauth2] Deal with opening and closing of browser through the network authentication handler

### PR DESCRIPTION
**Pull request by nirvn**
## Description

The goal here is to allow QGIS-based apps such as QField to implement their own UI to handle oauth2 authentication request.

It's pretty straight forward, mostly moving code around and creating functions in the network authentication handler & network access manager.

### CODE REVIEW

**Comment by nirvn** [comment]
@elpaso , any idea why the QGIS server fails here?

**Comment by elpaso** [comment]
> @elpaso , any idea why the QGIS server fails here?

Sorry I don't, @pblottiere is the architect of that test.

**Comment by pblottiere** [comment]
Hi @nirvn,

Actually the dataset for the WMS 1.3.0 testsuite is downloaded from http://cite.opengeospatial.org/teamengine/about/wms13/1.3.0/site/data-wms-1.3.0.zip but it seems offline today... It's not the first time so I'm going to embed the data in the `pyogctest` tool so it won't happen again.

**Comment by nirvn** [comment]
@pblottiere , thanks Paul.

**Comment by nirvn** [comment]
BTW, this PR also unlocks the possibility to have a nice in-app web panel opening instead of throwing people out of QGIS onto their browser. Not something I'm planning to do right away, merely a statement of what is possible.

File: src/core/network/qgsnetworkaccessmanager.h [hunk 1]
```diff
@@ -227,6 +227,20 @@ class CORE_EXPORT QgsNetworkAuthenticationHandler
      */
     virtual void handleAuthRequest( QNetworkReply *reply, QAuthenticator *auth );
 
+    /**
+     * Called to initiate a network authentication through external browser \a url.
+     *
```
Comment (nyalldawson) [inline src/core/network/qgsnetworkaccessmanager.h:232]: I think you should add a warning in these two methods that they aren't guaranteed to be called from the main thread, and may be called by a background thread.

**Comment by nirvn** [comment]
@nyalldawson , thanks for the review, comments addressed.

**Comment by nyalldawson** [comment]
LGTM

**Pull request merged**